src/Idris/Parser/Expr.hs

```diff
@@ -1288,16 +1290,26 @@ do_ :: SyntaxInfo -> IdrisParser PDo
 do_ syn
      = P.try (do keyword "let"
                  (i, ifc) <- withExtent name
-                 ty <- P.option Placeholder (do lchar ':'
-                                                expr' syn)
+                 ty' <- P.optional (do lchar ':'
+                                       expr' syn)
                  reservedOp "="
-                 (e, fc) <- withExtent $ expr syn
-                 return (DoLet fc RigW i ifc ty e))
+                 (e, fc) <- withExtent $ expr (syn { withAppAllowed = False })
+                 -- If there is an explicit type, this can’t be a pattern-matching let, so do not parse alternatives
```

Comment:
Should it be possible to say, for example, `let Nothing : Maybe (Int, String) = listToMaybe (reads "42") | Just _ => ...`?

Reply:
In the case of `let … in …` it *is* possible, but the type annotation is ignored, and allowing it not to be ignored is far beyond the scope of this PR, so I decided to disallow the combination here since it doesn’t add anything but looks confusing if the type happens to be wrong.

Reply:
What worries me is that `let` pattern matches on unary constructors, but since this is not something introduced by this PR, I will not block this.